Source/Core/Core/HLE/HLE_OS.cpp

```diff
@@ -204,76 +207,364 @@ void HLE_LogVFPrint(const Core::CPUThreadGuard& guard)
   HLE_LogFPrint(guard, ParameterType::VariableArgumentList);
 }
 
+namespace
+{
+class HLEPrintArgsVAList final : public HLEPrintArgs
+{
+public:
+  HLEPrintArgsVAList(const Core::CPUThreadGuard& guard, HLE::SystemVABI::VAList* va_list)
+      : m_guard(guard), m_va_list(va_list)
+  {
+  }
+
+  u32 GetU32() override { return m_va_list->GetArgT<u32>(); }
+  u64 GetU64() override { return m_va_list->GetArgT<u64>(); }
+  double GetF64() override { return m_va_list->GetArgT<double>(); }
+  std::string GetString(std::optional<u32> max_length) override
+  {
+    return max_length == 0u ? std::string() :
+                              PowerPC::MMU::HostGetString(m_guard, m_va_list->GetArgT<u32>(),
+                                                          max_length.value_or(0u));
+  }
+  std::u16string GetU16String(std::optional<u32> max_length) override
+  {
+    return max_length == 0u ? std::u16string() :
+                              PowerPC::MMU::HostGetU16String(m_guard, m_va_list->GetArgT<u32>(),
+                                                             max_length.value_or(0u));
+  }
+
+private:
+  const Core::CPUThreadGuard& m_guard;
+  HLE::SystemVABI::VAList* m_va_list;
+};
+}  // namespace
+
 std::string GetStringVA(Core::System& system, const Core::CPUThreadGuard& guard, u32 str_reg,
                         ParameterType parameter_type)
 {
   auto& ppc_state = system.GetPPCState();
 
-  std::string ArgumentBuffer;
-  std::string result;
   std::string string = PowerPC::MMU::HostGetString(guard, ppc_state.gpr[str_reg]);
-  auto ap =
+  std::unique_ptr<HLE::SystemVABI::VAList> ap =
       parameter_type == ParameterType::VariableArgumentList ?
-          std::make_unique<HLE::SystemVABI::VAListStruct>(system, guard,
-                                                          ppc_state.gpr[str_reg + 1]) :
-          std::make_unique<HLE::SystemVABI::VAList>(system, ppc_state.gpr[1] + 0x8, str_reg + 1);
+          std::make_unique<HLE::SystemVABI::VAListStruct>(guard, ppc_state.gpr[str_reg + 1]) :
+          std::make_unique<HLE::SystemVABI::VAList>(guard, ppc_state.gpr[1] + 0x8, str_reg + 1);
+
+  HLEPrintArgsVAList args(guard, ap.get());
+  return GetStringVA(&args, string);
+}
 
-  for (size_t i = 0; i < string.size(); i++)
+std::string GetStringVA(HLEPrintArgs* args, std::string_view string)
+{
+  std::string result;
+  for (size_t i = 0; i < string.size(); ++i)
   {
     if (string[i] == '%')
     {
-      ArgumentBuffer = '%';
-      i++;
-      if (string[i] == '%')
+      const size_t formatting_start_position = i;
+      ++i;
+      if (i < string.size() && string[i] == '%')
       {
         result += '%';
         continue;
       }
```

Comment:
I've removed some indents by inverting the conditions. I'm not sure what exactly you mean by removing the nested if/else?

Reply:
Sorry, I meant that by inverting the condition we can get rid of these `else` statements and remove these indents:
```c++
if (string[i] == '%') 
{
  // (...)
  if (i < string.size())
  {
    // (...)
  }
  else
  {
    return std::nullopt;
  }
  // etc.
}
else
{
  result += string[i];
}
```

So that's exactly what you did, thanks.